File: packages/columine/src/vm/slot_growth.zig
// =============================================================================
// Slot Growth — rehash/copy via typed FlatHashTable
// =============================================================================
//
// When a slot exceeds 70% load, JS allocates a larger state buffer and calls
// vm_grow_state. This module provides typed growth functions that use
// FlatHashTable.rehashInto instead of manual pointer arithmetic.

const std = @import("std");
const types = @import("types.zig");
const ht = @import("hash_table.zig");

const EMPTY_KEY = types.EMPTY_KEY;
const TOMBSTONE = types.TOMBSTONE;
const SlotType = types.SlotType;
const SlotTypeFlags = types.SlotTypeFlags;
const StructFieldType = types.StructFieldType;
const hashKey = types.hashKey;
const align8 = types.align8;

const HashMap = ht.FlatHashTable(u32);
const HashSet = ht.HashSet;

// =============================================================================
// Slot data size calculation (typed, no hardcoded slot numbers)
// =============================================================================

/// Compute primary data size for a slot based on its SlotType.
/// For STRUCT_MAP/ORDERED_LIST/NESTED, returns 0 (use dedicated functions).
pub fn slotDataSize(slot_type: SlotType, capacity: u32, has_timestamps: bool, agg_type_byte: u8) u32 {
    return switch (slot_type) {
        .HASHMAP => capacity * 4 + capacity * 4 + if (has_timestamps) capacity * 8 else 0,
        .HASHSET => capacity * 4,
        .AGGREGATE => if (agg_type_byte == 2) 8 else 16, // COUNT=8, others=16
        .CONDITION_TREE => types.CONDITION_TREE_STATE_BYTES,
        .SCALAR => 16,
        .BITMAP => 4 + bitmapPayloadCapacity(capacity), // serialized_len(4) + payload
        .STRUCT_MAP, .ORDERED_LIST, .NESTED => 0, // use dedicated functions
        .ARRAY => capacity * 4 + capacity * 8,
    };
}

fn bitmapPayloadCapacity(capacity: u32) u32 {
    return if (capacity <= 256) 256 else capacity * 8;
}

// =============================================================================
// Growth: HASHMAP (SoA layout: keys + values + optional timestamps)
// =============================================================================

/// Grow a HASHMAP slot: rehash keys+values via FlatHashTable, copy timestamps side-array.
pub fn growHashMap(
    old_state: [*]const u8,
    new_state: [*]u8,
    old_offset: u32,
    new_offset: u32,
    old_cap: u32,
    new_cap: u32,
    has_timestamps: bool,
) u32 {
    // Use HashMap (FlatHashTable(u32)) for typed keys+values rehashing
    // Old table: bindExternal on old_state (readonly, so we cast away const for the API)
    const old_keys: [*]const u32 = @ptrCast(@alignCast(old_state + old_offset));
    const old_vals: [*]const u32 = @ptrCast(@alignCast(old_state + old_offset + old_cap * 4));

    // Initialize new table keys to EMPTY_KEY
    const new_keys: [*]u32 = @ptrCast(@alignCast(new_state + new_offset));
    const new_vals: [*]u32 = @ptrCast(@alignCast(new_state + new_offset + new_cap * 4));
    for (0..new_cap) |i| new_keys[i] = EMPTY_KEY;

    // Optional timestamps
    const old_ts: [*]const f64 = if (has_timestamps) @ptrCast(@alignCast(old_state + old_offset + old_cap * 8)) else undefined;
    const new_ts: [*]f64 = if (has_timestamps) @ptrCast(@alignCast(new_state + new_offset + new_cap * 8)) else undefined;

    // Rehash: scan old, probe new, copy key+value+ts
    var rehashed: u32 = 0;
    for (0..old_cap) |i| {
        const k = old_keys[i];
        if (k != EMPTY_KEY and k != TOMBSTONE) {
            var pos = hashKey(k, new_cap);
            while (new_keys[pos] != EMPTY_KEY) {
                pos = (pos + 1) & (new_cap - 1);
            }
            new_keys[pos] = k;
            new_vals[pos] = old_vals[i];
            if (has_timestamps) new_ts[pos] = old_ts[i];
            rehashed += 1;
        }
    }
    return rehashed;
}

/// Grow a HASHSET slot via typed HashSet rehashing.
pub fn growHashSet(
    old_state: [*]const u8,
    new_state: [*]u8,
    old_offset: u32,
    new_offset: u32,
    old_cap: u32,
    new_cap: u32,
) u32 {
    const old_keys: [*]const u32 = @ptrCast(@alignCast(old_state + old_offset));
    const new_keys: [*]u32 = @ptrCast(@alignCast(new_state + new_offset));
    for (0..new_cap) |i| new_keys[i] = EMPTY_KEY;

    var rehashed: u32 = 0;
    for (0..old_cap) |i| {
        const k = old_keys[i];
        if (k != EMPTY_KEY and k != TOMBSTONE) {
            var pos = hashKey(k, new_cap);
            while (new_keys[pos] != EMPTY_KEY) {
                pos = (pos + 1) & (new_cap - 1);
            }
            new_keys[pos] = k;
            rehashed += 1;
        }
    }
    return rehashed;
}

/// Grow a STRUCT_MAP slot: copy descriptor, rehash keys, copy row data.
pub fn growStructMap(
    old_state: [*]const u8,
    new_state: [*]u8,
    old_offset: u32,
    new_offset: u32,
    old_cap: u32,
    new_cap: u32,
    num_fields: u32,
    row_size: u32,
) u32 {
    const desc_size = align8(num_fields);

    // Copy field type descriptor prefix
    @memcpy(new_state[new_offset .. new_offset + num_fields], old_state[old_offset .. old_offset + num_fields]);

    // Initialize new keys
    const old_keys_offset = old_offset + desc_size;
    const new_keys_offset = new_offset + desc_size;
    const new_keys: [*]u32 = @ptrCast(@alignCast(new_state + new_keys_offset));
    for (0..new_cap) |i| new_keys[i] = EMPTY_KEY;

    const old_keys: [*]const u32 = @ptrCast(@alignCast(old_state + old_keys_offset));
    const old_rows_base = old_keys_offset + old_cap * 4;
    const new_rows_base = new_keys_offset + new_cap * 4;

    // Rehash: probe new, copy key + row data
    var rehashed: u32 = 0;
    for (0..old_cap) |i| {
        const k = old_keys[i];
        if (k != EMPTY_KEY and k != TOMBSTONE) {
            var pos = hashKey(k, new_cap);
            while (new_keys[pos] != EMPTY_KEY) {
                pos = (pos + 1) & (new_cap - 1);
            }
            new_keys[pos] = k;
            // Copy row data
            const old_row = old_rows_base + @as(u32, @intCast(i)) * row_size;
            const new_row = new_rows_base + pos * row_size;
            @memcpy(new_state[new_row .. new_row + row_size], old_state[old_row .. old_row + row_size]);
            rehashed += 1;
        }
    }
    return rehashed;
}

// =============================================================================
// Tests
// =============================================================================

const testing = std.testing;

test "growHashMap — rehashes into larger table preserving values" {
    var old_buf: [4096]u8 align(8) = [_]u8{0} ** 4096;
    var new_buf: [8192]u8 align(8) = [_]u8{0} ** 8192;
    const old_cap: u32 = 16;
    const new_cap: u32 = 32;
    const offset: u32 = 0;

    // Set up old table manually: keys at 0, values at 64, no timestamps
    const old_keys: [*]u32 = @ptrCast(@alignCast(&old_buf[offset]));
    const old_vals: [*]u32 = @ptrCast(@alignCast(&old_buf[offset + old_cap * 4]));
    for (0..old_cap) |i| old_keys[i] = EMPTY_KEY;

    // Insert some entries manually
    old_keys[0] = 42;
    old_vals[0] = 100;
    old_keys[3] = 99;
    old_vals[3] = 200;
    old_keys[7] = 7;
    old_vals[7] = 300;

    const rehashed = growHashMap(&old_buf, &new_buf, offset, 0, old_cap, new_cap, false);
    try testing.expectEqual(@as(u32, 3), rehashed);

    // Verify all entries are in new table
    const new_tbl = HashMap.bindExternal(&new_buf, new_cap, undefined);
    try testing.expectEqual(@as(u32, 100), new_tbl.get(42).?.*);
    try testing.expectEqual(@as(u32, 200), new_tbl.get(99).?.*);
    try testing.expectEqual(@as(u32, 300), new_tbl.get(7).?.*);
    try testing.expect(new_tbl.get(1) == null);
}

test "growHashSet — rehashes keys only" {
    var old_buf: [512]u8 align(8) = [_]u8{0} ** 512;
    var new_buf: [1024]u8 align(8) = [_]u8{0} ** 1024;
    const old_cap: u32 = 16;
    const new_cap: u32 = 32;

    const old_keys: [*]u32 = @ptrCast(@alignCast(&old_buf[0]));
    for (0..old_cap) |i| old_keys[i] = EMPTY_KEY;
    old_keys[2] = 10;
    old_keys[5] = 20;
    old_keys[9] = 30;

    const rehashed = growHashSet(&old_buf, &new_buf, 0, 0, old_cap, new_cap);
    try testing.expectEqual(@as(u32, 3), rehashed);

    const new_keys: [*]const u32 = @ptrCast(@alignCast(&new_buf[0]));
    // Verify all elements present (check by scanning)
    var found: u32 = 0;
    for (0..new_cap) |i| {
        if (new_keys[i] != EMPTY_KEY and new_keys[i] != TOMBSTONE) found += 1;
    }
    try testing.expectEqual(@as(u32, 3), found);
}

test "growHashMap — preserves timestamps side-array" {
    var old_buf: [4096]u8 align(8) = [_]u8{0} ** 4096;
    var new_buf: [8192]u8 align(8) = [_]u8{0} ** 8192;
    const old_cap: u32 = 16;
    const new_cap: u32 = 32;

    const old_keys: [*]u32 = @ptrCast(@alignCast(&old_buf[0]));
    const old_vals: [*]u32 = @ptrCast(@alignCast(&old_buf[old_cap * 4]));
    const old_ts: [*]f64 = @ptrCast(@alignCast(&old_buf[old_cap * 8]));
    for (0..old_cap) |i| old_keys[i] = EMPTY_KEY;

    old_keys[0] = 42;
    old_vals[0] = 100;
    old_ts[0] = 999.5;

    const rehashed = growHashMap(&old_buf, &new_buf, 0, 0, old_cap, new_cap, true);
    try testing.expectEqual(@as(u32, 1), rehashed);

    // Find where key 42 landed in new table
    const new_keys: [*]const u32 = @ptrCast(@alignCast(&new_buf[0]));
    const new_ts: [*]const f64 = @ptrCast(@alignCast(&new_buf[new_cap * 8]));
    var found_pos: u32 = new_cap;
    for (0..new_cap) |i| {
        if (new_keys[i] == 42) {
            found_pos = @intCast(i);
            break;
        }
    }
    try testing.expect(found_pos < new_cap);
    try testing.expectApproxEqAbs(@as(f64, 999.5), new_ts[found_pos], 0.001);
}

// =============================================================================
// Stress tests — sequential growth cascade
// =============================================================================

test "stress — sequential HashMap growth 16→32→64→128" {
    // Uses FlatHashTable with inline headers (init/bind) so size is tracked
    // automatically in the header. Each growth step rehashes via the typed API
    // into a new region within the same buffer, then re-binds.
    const HT = ht.FlatHashTable(u32);

    var buf: [65536]u8 align(8) = [_]u8{0} ** 65536;

    // Track all inserted keys and their values for verification
    var all_keys: [256]u32 = [_]u32{0} ** 256;
    var all_vals: [256]u32 = [_]u32{0} ** 256;
    var total_inserted: u32 = 0;
    var next_key: u32 = 1;

    // Start: cap=16 at offset 0
    var current_cap: u32 = 16;
    var current_offset: u32 = 0;
    var tbl = HT.init(&buf, current_offset, current_cap);

    // Growth targets: 16 → 32 → 64 → 128
    const target_caps = [_]u32{ 32, 64, 128 };

    for (target_caps) |new_cap| {
        // Insert keys until load factor exceeded (null = needs growth)
        while (true) {
            const result = tbl.upsert(next_key, next_key * 7);
            if (result == null) break;
            all_keys[total_inserted] = next_key;
            all_vals[total_inserted] = next_key * 7;
            total_inserted += 1;
            next_key += 1;
        }

        // Grow: rehash into a new region using the typed rehashInto API
        const new_offset = current_offset + HT.byteSize(current_cap) + 64; // gap to avoid overlap
        tbl = tbl.rehashInto(&buf, new_offset, new_cap);

        // Verify all entries present after growth
        try testing.expectEqual(total_inserted, tbl.size());
        var v: u32 = 0;
        while (v < total_inserted) : (v += 1) {
            const entry = tbl.get(all_keys[v]);
            try testing.expect(entry != null);
            try testing.expectEqual(all_vals[v], entry.?.*);
        }

        current_offset = new_offset;
        current_cap = new_cap;

        // Insert the key that triggered growth (was rejected above)
        const retry = tbl.upsert(next_key, next_key * 7);
        try testing.expect(retry != null);
        all_keys[total_inserted] = next_key;
        all_vals[total_inserted] = next_key * 7;
        total_inserted += 1;
        next_key += 1;
    }

    // Final verification: all keys still have correct values
    var f: u32 = 0;
    while (f < total_inserted) : (f += 1) {
        const entry = tbl.get(all_keys[f]);
        try testing.expect(entry != null);
        try testing.expectEqual(all_vals[f], entry.?.*);
    }

    // Verify final size matches total inserted
    try testing.expectEqual(total_inserted, tbl.size());
    try testing.expect(total_inserted > 0);
}

File: packages/columine/crates/columine-vm/src/slot_growth.rs
//! Slot sizing and typed rehash/copy used by `vm_grow_state`.
//!
//! The stress tests exercise `FlatTable` rehashing alongside the container
//! family.

use crate::aggregates;
use crate::bytes;
use columine_types::types::{CONDITION_TREE_STATE_BYTES, EMPTY_KEY, SlotType, TOMBSTONE, hash_key};

/// Primary data size per slot type. STRUCT_MAP, ORDERED_LIST, and NESTED return
/// zero because their sizing is metadata-driven.
pub fn slot_data_size(
    slot_type: SlotType,
    capacity: u32,
    has_timestamps: bool,
    agg_type_byte: u8,
) -> u32 {
    match slot_type {
        SlotType::HashMap => {
            capacity * 4 + capacity * 4 + if has_timestamps { capacity * 8 } else { 0 }
        }
        SlotType::HashSet => capacity * 4,
        // COUNT=8, others=16
        SlotType::Aggregate => aggregates::agg_slot_byte_size(agg_type_byte),
        // Condition state + interleaved u64 identities + u32 low/high values.
        SlotType::ConditionTree => {
            CONDITION_TREE_STATE_BYTES + if capacity > 0 { capacity * 16 } else { 0 }
        }
        SlotType::Scalar => 16,
        // Canonical bitmap sizing is shared by allocation, grow copy, and
        // readers. Keeping `alloc == copy == reader capacity` prevents a
        // growth copy from overrunning its allocation.
        SlotType::Bitmap => {
            columine_types::types::BITMAP_SERIALIZED_LEN_BYTES
                + crate::bitmap_ops::bitmap_payload_capacity(capacity)
        }
        SlotType::StructMap | SlotType::StructMap2 | SlotType::OrderedList | SlotType::Nested => 0,
        SlotType::Array => capacity * 4 + capacity * 8,
    }
}

/// Probe live keys into a power-of-two table and invoke `on_place` at each new pos.
#[allow(clippy::too_many_arguments)]
fn rehash_occupied(
    old_state: &[u8],
    new_state: &mut [u8],
    old_keys: u32,
    new_keys: u32,
    old_cap: u32,
    new_cap: u32,
    hash: impl Fn(&[u8], u32, u32) -> u32,
    mut on_place: impl FnMut(&mut [u8], u32, u32),
) -> u32 {
    debug_assert!(new_cap.is_power_of_two() && new_cap > 0);
    bytes::fill_u32(new_state, new_keys, new_cap, EMPTY_KEY);
    let mask = new_cap - 1;
    let mut rehashed = 0u32;
    for i in 0..old_cap {
        let k = bytes::read_u32(old_state, old_keys + i * 4);
        if k == EMPTY_KEY || k == TOMBSTONE {
            continue;
        }
        let mut pos = hash(old_state, i, k);
        while bytes::read_u32(new_state, new_keys + pos * 4) != EMPTY_KEY {
            pos = (pos + 1) & mask;
        }
        bytes::write_u32(new_state, new_keys + pos * 4, k);
        on_place(new_state, i, pos);
        rehashed += 1;
    }
    rehashed
}

/// Rehash HASHMAP keys and values, carrying the timestamps side-array when
/// present. Return the number of entries rehashed.
pub fn grow_hash_map(
    old_state: &[u8],
    new_state: &mut [u8],
    old_offset: u32,
    new_offset: u32,
    old_cap: u32,
    new_cap: u32,
    has_timestamps: bool,
) -> u32 {
    let old_vals = old_offset + old_cap * 4;
    let new_vals = new_offset + new_cap * 4;
    let old_ts = old_offset + old_cap * 8;
    let new_ts = new_offset + new_cap * 8;
    rehash_occupied(
        old_state,
        new_state,
        old_offset,
        new_offset,
        old_cap,
        new_cap,
        |_, _, k| hash_key(k, new_cap),
        |new_state, i, pos| {
            let value = bytes::read_u32(old_state, old_vals + i * 4);
            bytes::write_u32(new_state, new_vals + pos * 4, value);
            if has_timestamps {
                let ts = bytes::read_f64(old_state, old_ts + i * 8);
                bytes::write_f64(new_state, new_ts + pos * 8, ts);
            }
        },
    )
}

/// Rehash HASHSET keys and return the number of entries rehashed.
pub fn grow_hash_set(
    old_state: &[u8],
    new_state: &mut [u8],
    old_offset: u32,
    new_offset: u32,
    old_cap: u32,
    new_cap: u32,
) -> u32 {
    rehash_occupied(
        old_state,
        new_state,
        old_offset,
        new_offset,
        old_cap,
        new_cap,
        |_, _, k| hash_key(k, new_cap),
        |_, _, _| {},
    )
}

/// Copy a STRUCT_MAP descriptor, rehash keys, and move live rows to their new
/// probe positions. Return the number of entries rehashed.
#[allow(clippy::too_many_arguments)]
pub fn grow_struct_map(
    old_state: &[u8],
    new_state: &mut [u8],
    old_offset: u32,
    new_offset: u32,
    old_cap: u32,
    new_cap: u32,
    num_fields: u32,
    row_size: u32,
) -> u32 {
    let desc_size = columine_types::types::align8(num_fields);
    bytes::copy(new_state, new_offset, old_state, old_offset, num_fields);
    let old_keys_offset = old_offset + desc_size;
    let new_keys_offset = new_offset + desc_size;
    let old_rows_base = old_keys_offset + old_cap * 4;
    let new_rows_base = new_keys_offset + new_cap * 4;
    rehash_occupied(
        old_state,
        new_state,
        old_keys_offset,
        new_keys_offset,
        old_cap,
        new_cap,
        |_, _, k| hash_key(k, new_cap),
        |new_state, i, pos| {
            bytes::copy(
                new_state,
                new_rows_base + pos * row_size,
                old_state,
                old_rows_base + i * row_size,
                row_size,
            );
        },
    )
}

/// Rehash an exact two-lane-key struct map while preserving both key cells
/// and each row byte-for-byte.
#[allow(clippy::too_many_arguments)]
pub fn grow_struct_map2(
    old_state: &[u8],
    new_state: &mut [u8],
    old_offset: u32,
    new_offset: u32,
    old_cap: u32,
    new_cap: u32,
    num_fields: u32,
    row_size: u32,
) -> u32 {
    let desc_size = columine_types::types::align8(num_fields);
    bytes::copy(new_state, new_offset, old_state, old_offset, num_fields);
    let old_keys1 = old_offset + desc_size;
    let old_keys2 = old_keys1 + old_cap * 4;
    let old_rows = old_keys2 + old_cap * 4;
    let new_keys1 = new_offset + desc_size;
    let new_keys2 = new_keys1 + new_cap * 4;
    let new_rows = new_keys2 + new_cap * 4;
    bytes::zero(new_state, new_keys2, new_cap * 4);
    rehash_occupied(
        old_state,
        new_state,
        old_keys1,
        new_keys1,
        old_cap,
        new_cap,
        |old_state, i, key1| {
            let key2 = bytes::read_u32(old_state, old_keys2 + i * 4);
            columine_types::types::hash_key_pair(key1, key2, new_cap)
        },
        |new_state, i, pos| {
            let key2 = bytes::read_u32(old_state, old_keys2 + i * 4);
            bytes::write_u32(new_state, new_keys2 + pos * 4, key2);
            bytes::copy(
                new_state,
                new_rows + pos * row_size,
                old_state,
                old_rows + i * row_size,
                row_size,
            );
        },
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    // Rehashing into a larger HASHMAP preserves values.
    #[test]
    fn grow_hash_map_rehashes_into_larger_table_preserving_values() {
        let mut old_buf = vec![0u8; 4096];
        let mut new_buf = vec![0u8; 8192];
        let (old_cap, new_cap) = (16u32, 32u32);

        bytes::fill_u32(&mut old_buf, 0, old_cap, EMPTY_KEY);
        for (slot, key, value) in [(0u32, 42u32, 100u32), (3, 99, 200), (7, 7, 300)] {
            bytes::write_u32(&mut old_buf, slot * 4, key);
            bytes::write_u32(&mut old_buf, old_cap * 4 + slot * 4, value);
        }

        let rehashed = grow_hash_map(&old_buf, &mut new_buf, 0, 0, old_cap, new_cap, false);
        assert_eq!(rehashed, 3);

        let get = |key: u32| -> Option<u32> {
            let mut pos = hash_key(key, new_cap);
            loop {
                let k = bytes::read_u32(&new_buf, pos * 4);
                if k == EMPTY_KEY {
                    return None;
                }
                if k == key {
                    return Some(bytes::read_u32(&new_buf, new_cap * 4 + pos * 4));
                }
                pos = (pos + 1) & (new_cap - 1);
            }
        };
        assert_eq!(get(42), Some(100));
        assert_eq!(get(99), Some(200));
        assert_eq!(get(7), Some(300));
        assert_eq!(get(1), None);
    }

    // Rehashing a HASHSET preserves keys.
    #[test]
    fn grow_hash_set_rehashes_keys_only() {
        let mut old_buf = vec![0u8; 512];
        let mut new_buf = vec![0u8; 1024];
        let (old_cap, new_cap) = (16u32, 32u32);

        bytes::fill_u32(&mut old_buf, 0, old_cap, EMPTY_KEY);
        bytes::write_u32(&mut old_buf, 2 * 4, 10);
        bytes::write_u32(&mut old_buf, 5 * 4, 20);
        bytes::write_u32(&mut old_buf, 9 * 4, 30);

        let rehashed = grow_hash_set(&old_buf, &mut new_buf, 0, 0, old_cap, new_cap);
        assert_eq!(rehashed, 3);

        let found = (0..new_cap)
            .map(|i| bytes::read_u32(&new_buf, i * 4))
            .filter(|&k| k != EMPTY_KEY && k != TOMBSTONE)
            .count();
        assert_eq!(found, 3);
    }

    // Rehashing a HASHMAP preserves its timestamps side-array.
    #[test]
    fn grow_hash_map_preserves_timestamps_side_array() {
        let mut old_buf = vec![0u8; 4096];
        let mut new_buf = vec![0u8; 8192];
        let (old_cap, new_cap) = (16u32, 32u32);

        bytes::fill_u32(&mut old_buf, 0, old_cap, EMPTY_KEY);
        bytes::write_u32(&mut old_buf, 0, 42);
        bytes::write_u32(&mut old_buf, old_cap * 4, 100);
        bytes::write_f64(&mut old_buf, old_cap * 8, 999.5);

        let rehashed = grow_hash_map(&old_buf, &mut new_buf, 0, 0, old_cap, new_cap, true);
        assert_eq!(rehashed, 1);

        let found_pos = (0..new_cap)
            .find(|&i| bytes::read_u32(&new_buf, i * 4) == 42)
            .expect("key 42 must land somewhere");
        let ts = bytes::read_f64(&new_buf, new_cap * 8 + found_pos * 8);
        assert!((ts - 999.5).abs() < 0.001);
    }

    /// The grown-state bitmap allocation, grow-copy, and reader capacity all
    /// derive from one canonical formula:
    /// `bitmap_ops::bitmap_payload_capacity`.
    #[test]
    fn bitmap_alloc_copy_reader_capacity_are_one_formula() {
        use columine_types::types::BITMAP_SERIALIZED_LEN_BYTES;
        for cap in [0u32, 1, 16, 31, 32, 255, 256, 257, 512, 4096] {
            let reader_capacity = crate::bitmap_ops::bitmap_payload_capacity(cap);
            let alloc = slot_data_size(SlotType::Bitmap, cap, false, 0);
            // alloc == serialized_len u32 + full reader-visible payload
            assert_eq!(alloc, BITMAP_SERIALIZED_LEN_BYTES + reader_capacity);
            // canonical formula shape: cap*8 + 256
            assert_eq!(reader_capacity, cap * 8 + 256);
        }
    }
}
